sys/androidmedia/gstamc.c

```diff
@@ -709,10 +738,106 @@ gst_amc_codec_enable_adaptive_playback (GstAmcCodec * codec,
 
   adaptive_supported = gst_amc_codec_is_feature_supported (codec, format,
       "adaptive-playback");
+
   if (adaptive_supported) {
+    JNIEnv *env = gst_jni_get_env ();
+    /* if unlikely we cannot retrieve the maximum frame size supported,
+     * the following 4K UHD values will be set */
+    jint max_height = 2160;
+    jint max_width = 3840;
+
+    if (!media_codec_info.video_caps.klass) {
+      GST_ERROR ("Video caps not supported, requires API 21");
+    } else {
+      jobject codec_info;
+      jobject capabilities;
+      jobject video_caps = NULL;
+      jstring jtmpstr = NULL;
+      jboolean supported;
+
+      J_CALL_OBJ (codec_info /* = */ , codec->object,
+          media_codec.get_codec_info);
+
+      AMC_CHK (gst_amc_format_get_jstring (format, "mime", &jtmpstr));
+      J_CALL_OBJ (capabilities /* = */ , codec_info,
+          media_codec_info.get_capabilities_for_type, jtmpstr);
+      J_DELETE_LOCAL_REF (jtmpstr);
+
+      J_CALL_OBJ (video_caps /* = */ , capabilities,
+          codec_capabilities.get_video_caps);
+
+      /* NOTE: We tried getSupportedHeights and getSupportedWidthsFor
+       *  but we obtained non standard resolutions like 1072x8688.
+       *  So we implemented the other way below to obtain the maximum standard
+       *  size supported, trying 8K, then 4K DCI, then 4K UHD and using FHD
+       *  otherwise.
+       *  For now we keep this unneeded code because it was tricky to have it
+       *  working (Range is a template class) and we want the values reported
+       *  logged for every board. It may be safely deleted if we finally find
+       *  it unnecessary. */
+      {
+        jobject heights;
+        jobject widths;
+        jobject upper;
```

Comment:
this objects must be freed with J_DELETE_LOCAL_REF() after usage.

Reply:
there's a pattern to do that: just initialize all the jobject vars to NULL, and put J_DELETE_LOCAL_REF () with each of them after "error" label